draft-ietf-tls-tls13.md

```diff
@@ -1953,11 +1947,14 @@ The "extension_data" field of this extension contains a
 %%% Key Exchange Messages
 
        struct {
-           select (role) {
-               case client:
+           select (Handshake.msg_type) {
+               case client_hello:
                    KeyShareEntry client_shares<0..2^16-1>;
 
-               case server:
+               case hello_retry_request:
+                   NamedGroup selected_group;
+
+               case server_hello:
                    KeyShareEntry server_share;
            }
        } KeyShare;
```

Comment:
Honestly, the key_share / supported_groups split is kinda silly. If we didn't have the TLS 1.2 constraint, I imagine we would have made it one extension and just left the missing key shares blank (or maybe did something tighter).

I used key_share because it makes it nicely symmetric with the ServerHello one (if the server sends key_share in either message, it means the server wants to do (EC)DHE for *this* connection). Whereas the server might send supported_groups in either PSK-only or (EC)DHE-ful handshakes.

This also lets the description for KeyShare processing stay in one place. The full story is told in the description of the key_share extension.